### ISSUE TITLE
Utilize CraftOS ROMs via .jar's from OG ComputerCraft or CC:T

### ISSUE THREAD

**Issue by Psypherix** [MCJack123/craftos2#191]
**Is your feature request related to a problem? Please describe.**
Not a problem with the Emulator, but just a feature I've always wanted in a ComputerCraft emulator since I've been developing for it. When I write ComputerCraft Programs, I try to test them against multiple versions of the mod to ensure compatibility with as many versions as I can make it or to at least get it to run on the targeted versions of Minecraft I want it to run on which determine the versions of the mod I have to target as well.

**Describe the solution you'd like**
This might be outside the scope of the vision of this project, But I think it can be rolled out in 2 phases:

**Phase 1**: Allow CraftOS-PC to use the LUA ROM that is contained from within a ComputerCraft .jar file. Limited to a select few versions at first in order to get this part figured out with having the mod use an external rom folder. 

**Phase 2**: Have a folder where the user can store the various .jars from previous versions of ComputerCraft and previous versions of CC:T, and have some type of Selector Window that allows the user to select which version of CraftOS they wish to emulate.

A little on how I think this could be achieved:
Jar files are essentially ZIP Files, meaning that their data can be read easily. We could easily navigate the contents of the Jar file, and either use it directly from the Jar or extract the ROM located in the mod's 'lua' folder into a folder labeled with the version of CraftOS and then delete the .Jar once we're done.

I also realize that some tweaks have been made to ComputerCraft's Java code over time, and because of that, some of the LUA ROM code for those versions of the mod also differ. An example of this was with the introduction of turtles and pocket computers where the LUA ROM code changed to utilize these changes. This means that the Emulator might have to do some things a little differently in order to be compatible with the chosen version of the ROM the user wishes to emulate. My suggestion is to have the Emulator look inside the Jar to determine what version of the mod is running by utilizing the MCMod.info so that it knows what changes to make in order run correctly and then having the emulator make the appropriate changes to it's emulation environment to have that version of the LUA ROM run correctly.

**Describe alternatives you've considered**
There currently are no other emulators that do this or at least do it the way I mentioned.

**Additional context**
I've been wishing for an emulator to do this and do it elegantly. I really feel like this will truly make this emulator one of the best ones out there with that feature addition. I'd love to help develop this with you if you'd allow me.

**Comment by SkyTheCodeMaster** [comment]
I believe you can already change the ROM that CraftOS-PC uses, so you can take the ROM out of the jar and have Lua use that instead of the built in one. (I believe on windows, the local is `%APPDATA%/Local/CraftOS-PC`, but take it with a grain of salt.)

**Comment by MCJack123** [comment]
Most of the changes made to ComputerCraft happen in the core, and not specifically the ROM. The only modifications I see made to CC:T's ROM are either new modules, or tweaks to pre-existing programs, though there have been a number of new API functions that pop up (like `colors.toBlit` or `textutils.unserializeJSON`).

You can use your own ROM directory using the `--rom` flag on the command line. If you want to use the ROM from, say, 1.89.2, you can extract `assets/computercraft/lua` from the JAR file and run CraftOS-PC with the `--rom` flag pointing to the directory.

As for having a window to do this: SDL does not provide any user interface libraries. Implementing any UI outside CraftOS would either a) have to be developed from scratch, b) require loading some prebuilt library like wxWidgets, GTK, or Qt, or c) require platform-specific code to set up the UI. None of these solutions are very good, and this is why things like the release notes viewer are implemented inside a separate computer rather than setting up a proper UI.

I don't imagine many users will need this sort of functionality, as they'll be writing programs for their own use which won't involve other versions. There are definitely use cases for this sort of thing, but having it be a core feature isn't really important enough to involve building a whole system to not only load a ComputerCraft JAR, but also read the ROM from it (which would involve setting up JAR mounts - thankfully, the VFS interface could help with that), parse the mod info, and determine from that what to add, remove or tweak. It's quite a bit more than I'd want to add in, especially if the chances of it being used are very low. (Heck, many users don't even know about the screenshot functionality.)

A lot of the involved functionality can be replicated from the Lua side, since you have full control over what globals are available from any script. My recommendation would be to write a Lua script (that you could make the startup file, or even add to the BIOS (or make the BIOS this script and set up your own chainloader!)) that takes a ComputerCraft version to target, and deletes or modifies whatever core functions that are different. This could be accomplished by setting up a table of versions with the features added, and removing anything that is added before the target:
```lua
local versionDeltas = {
    ["1.95.2"] = {
        {"fs", "attributes", replace = function(original) return function(path) local attr = original(path) attr.isReadOnly = false return attr end end}
    },
    ["1.92.0"] = {
        {"fs", "combine", replace = function(original) return function(a, b) return original(a, b) end end}
    },
    -- etc.
}

write("Enter the target version: ")
local target = read()
for k,v in pairs(versionDeltas) do if k > target then
    for _, item in ipairs(v) do
        _G[item[1]][item[2]] = item.replace and item.replace(_G[item[1]][item[2]]) or nil
    end
end
```
This could then be combined with changing the ROM path (possibly through a shell script) to get the required compatibility set.

Other options include using older versions of CraftOS-PC, or using CCEmuX (which is able to load older JAR files entirely) for compatibility testing. (If you're aiming for complete compatibility, using CCEmuX for testing purposes may be better as it has guaranteed compatibility with ComputerCraft. Developing in CraftOS-PC and testing for compatibility in CCEmuX is a good way to make sure your program works on everything.)

If you do want the functionality coming from the emulator, you could write a plugin that handles this all behind the scenes, manipulating the mounts to point to the right versions and changing the functions to work like the version required, and displaying a window like you said if you want. The API should make the interaction between plugin and emulator easy enough, but you'll need to know a decent amount about C++ to accomplish this.

I might implement loading ROMs from JARs, as that would not be too much since Poco already has a ZIP library that would be drag-and-drop to integrate, and there's already functionality for loading data (including ROMs) from virtual sources. However, I'm not going to have it try to adapt to be able to load it - especially since the ROM hasn't depended on the core for a long time.

TL;DR Cool idea, but it would take too much work to do and not many users would need it; and the functionality can be implemented by the users if desired, through Lua or C++.
